### PR TITLE
Add `createFontStack` for metrics-based font fallbacks

### PR DESCRIPTION
**Pull request by michaeltaranto**
Creates metrics-based `@font-face` declarations to improve the alignment of font family fallbacks, which can dramatically improve the [Cumulative Layout Shift](https://web.dev/cls/) metric for sites that depend on a web font.

### Example usage

```ts
import { createFontStack } from '@capsizecss/core';
import lobster from '@capsizecss/metrics/lobster';
import helveticaNeue from '@capsizecss/metrics/helveticaNeue';
import arial from '@capsizecss/metrics/arial';

const { fontFamily, fontFaces } = createFontStack([
  lobster,
  helveticaNeue,
  arial,
]);
```

The returned values are the computed font family and the generated font face declarations:
```ts
// `fontFamily`
"Lobster, 'Lobster Fallback: Helvetica Neue', 'Lobster Fallback: Arial'"
```

```css
/* `fontFaces` */
@font-face {
  font-family: 'Lobster Fallback: Helvetica Neue';
  src: local('Helvetica Neue');
  ascent-override: 115.1741%;
  descent-override: 28.7935%;
  size-adjust: 86.8251%;
}
@font-face {
  font-family: 'Lobster Fallback: Arial';
  src: local('Arial');
  ascent-override: 113.5679%;
  descent-override: 28.392%;
  size-adjust: 88.053%;
}
```

### CODE REVIEW

**Review by mrm007 — COMMENTED** [review]
Really excited for this work to finally get published! 👏👏👏

Just a few nitpicky comments from me 🏃‍♂️

File: README.md [hunk 1]
```diff
@@ -163,21 +164,157 @@ This metadata is extracted from the metrics tables inside the font itself. There
...
+</style>
+```
+
+This will produce the following CSS:
+
+```css
+.heading {
+  font-family: Lobster, 'Lobster Fallback: Helvetica Neue',
+    'Lobster Fallback: Arial';
+}
```
[the remark this answers is not in the document]
Comment (michaeltaranto) [reply README.md:223]: Not sure about opting out for single lines without have to have a visible comment. But the CSS is still valid, so not sure its necessary.

File: packages/core/src/createFontStack.ts [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+}
+const calculateOverrideValues = ({
+  metrics,
+  fallbackMetrics,
+}: OverrideValuesParams): AtRule.FontFace => {
+  // Calculate size adjust
+  let sizeAdjust = 1;
+  if (
+    typeof metrics.xWidthAvg === 'number' &&
+    typeof fallbackMetrics.xWidthAvg === 'number'
```
[the remark this answers is not in the document]
Comment (michaeltaranto) [reply packages/core/src/createFontStack.ts:41]: Yeah good question, this was less about the type matching and more about ensuring we only apply a valid size-adjust. I think this can be improved.

File: packages/core/src/createFontStack.ts [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+      fontFace.push('}');
+
+      return fontFace.join('\n');
+    })
+    .join('\n');
+};
+
+export const quoteIfNeeded = (name: string) =>
+  !/^[a-zA-Z][^"'][a-zA-Z\d-]+[a-zA-Z\d]$/.test(name) ? `'${name}'` : name;
```
[the remark this answers is not in the document]
Comment (michaeltaranto) [reply packages/core/src/createFontStack.ts:107]: Adding more test cases and refactored the function to handle a few more edge cases.

File: packages/core/src/createFontStack.test.ts [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+      { input: `Goudy Bookletter 1911`, output: `'Goudy Bookletter 1911'` },
+      { input: `Red/Black`, output: `'Red/Black'` },
+      { input: `"Lucida" Grande`, output: `'"Lucida" Grande'` },
+      { input: `Ahem!`, output: `'Ahem!'` },
+      { input: `test@foo`, output: `'test@foo'` },
+      { input: `#POUND`, output: `'#POUND'` },
+      { input: `Hawaii 5-0`, output: `'Hawaii 5-0'` },
+      { input: `sans-serif`, output: `sans-serif` },
+      { input: `asdasdasd`, output: `asdasdasd` },
+      { input: `"Lucide Grande"`, output: `'"Lucide Grande"'` },
```
[the remark this answers is not in the document]
Comment (michaeltaranto) [reply packages/core/src/createFontStack.test.ts:59]: Have refactored the code and test cases in this space. So maybe re-review when pushed if you can.

File: packages/core/src/createFontStack.test.ts [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+      { input: `"Lucide`, output: `'"Lucide'` },
+      { input: `'Sasd asdasd'`, output: `''Sasd asdasd''` },
```
[the remark this answers is not in the document]
Comment (michaeltaranto) [reply packages/core/src/createFontStack.test.ts:61]: As above

File: packages/core/src/createFontStack.ts [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+  | 'src'
+  | 'fontFamily'
+  | 'ascentOverride'
+  | 'descentOverride'
+  | 'lineGapOverride'
+  | 'sizeAdjust'
+>;
+type CreateFontStackOptions = {
+  /** Additional properties to add to the generated `@font-face` declarations.
+   *
```
[the remark this answers is not in the document]
Comment (michaeltaranto) [reply packages/core/src/createFontStack.ts:120]: DIdnt have much success with that plugin, but i'll update these two instances manually.

File: packages/core/src/createFontStack.ts [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    ...optionsArg,
+  };
+  const { familyName } = metrics;
+
+  const fontFamilys: string[] = [quoteIfNeeded(familyName)];
+  const fontFaces: FontFace[] = [];
+
+  fallbacks.forEach((fallbackMetrics) => {
+    const fontFamily = `'${familyName} Fallback${
+      fallbacks.length > 1 ? `: ${fallbackMetrics.familyName}` : ''
```
[the remark this answers is not in the document]
Comment (michaeltaranto) [reply packages/core/src/createFontStack.ts:162]: Improved the handling here in along with the tests.

File: pnpm-lock.yaml [hunk 8]
```diff
@@ -5268,6 +5271,23 @@ packages:
       deepmerge: 4.2.2
       media-query-parser: 2.0.2
       outdent: 0.8.0
+    dev: false
+
+  /@vanilla-extract/css/1.9.2:
```
[the remark this answers is not in the document]
Comment (michaeltaranto) [reply pnpm-lock.yaml:5276]: Hmm, not sure why 🤔 . Removed the older one manually, and made pnpm fix the lock file which seems to have removed the older one successfully 👌

File: packages/core/src/createFontStack.ts [hunk 9]
```diff
@@ -0,0 +1,186 @@
+import { AtRule } from 'csstype';
+import { round } from './round';
+
+const formatValue = (value: number) => `${round(value * 100)}%`;
+
+export const toCssProperty = (property: string) =>
+  property.replace(/([A-Z])/g, (property) => `-${property.toLowerCase()}`);
+
+interface FontMetrics {
```
[the remark this answers is not in the document]
Comment (michaeltaranto) [reply packages/core/src/createFontStack.ts:9]: Yep, wanted to talk through the type interfaces. Let's chat through this one.

File: README.md [hunk 10]
```diff
@@ -163,21 +164,157 @@ This metadata is extracted from the metrics tables inside the font itself. There
...
+@font-face {
+  font-family: 'Lobster Fallback: Arial';
+  src: local('Arial');
+  ascent-override: 113.5679%;
+  descent-override: 28.392%;
+  size-adjust: 88.053%;
+}
+```
+
+#### Usage with CSS-in-JS framework
```
[the remark this answers is not in the document]
Comment (mrm007) [reply README.md]: Oh, I didn't realise these headings weren't in the TOC (they are now).

File: packages/core/src/createFontStack.ts [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+export const quoteIfNeeded = (name: string) => {
+  const quotedMatch = name.match(/^['"](?<name>.*)['"]$/);
+  if (quotedMatch && quotedMatch.groups?.name) {
+    // Escape double quotes in middle of name
+    return `"${quotedMatch.groups.name.split(`"`).join(`\"`)}"`;
+  }
+
+  if (/^"/.test(name)) {
+    // Complete double quotes if incomplete and escape double quotes in middle
+    const [, ...restName] = name;
```
Comment (mrm007) [inline packages/core/src/createFontStack.ts:81]: Does this work well with compound emoji? For example, I know `.split('')` has an unexpected behaviour:

```js
'👨‍👩‍👧‍👧'.split('') // => Array(11) [ "\ud83d", "\udc68", "‍", "\ud83d", "\udc69", "‍", "\ud83d", "\udc67", "‍", "\ud83d", … ]
```

Not sure if emojis are allowed in font names, so apologies if this is a very narrow edge case.

Comment (michaeltaranto) [reply packages/core/src/createFontStack.ts:81]: Not sure we need to worry about handling this level of complexity just yet. Given there is no impact on public API, would be happy to evolve support as required here.

**Comment by phil-lgr** [comment]
Really hype about this one, can we help in anyway? I'd like to try and test this

**Comment by michaeltaranto** [comment]
> Really hyped about this one, can we help in any way? I'd like to try and test this

Apologies for the festive season delays on this one 😄. Hoping to get it through soon. The full benefit will be had in tandem with the update to the `metrics` package which will be PR'd shortly.

FWIW, there is a snapshot deployment you can take for a spin:
```bash
npm install @capsizecss/core@0.0.0-create-font-stack-20221130045640
```
You can use the updated 
```bash
npm install @capsizecss/metrics@0.0.0-x-width-avg-20230123213508
```

**Comment by michaeltaranto** [comment]
> Really hyped about this one, can we help in any way? I'd like to try and test this

Apologies for the festive season delays on this one 😄. Hoping to get it through soon. The full benefit will be had in tandem with the update to the `metrics` package which will be PR'd shortly.

FWIW, there is a snapshot deployment of this PR you can take for a spin:
```bash
npm install @capsizecss/core@0.0.0-create-font-stack-20221130045640
```

To test out the fallbacks with the `size-adjust` property you can install the updated metrics package too:
```bash
npm install @capsizecss/metrics@0.0.0-x-width-avg-20230123213508
```

File: packages/core/src/createFontStack.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (michaeltaranto) [reply packages/core/src/createFontStack.ts]: Resolved by merging in [this PR](https://github.com/seek-oss/capsize/pull/122) and [this commit below](https://github.com/seek-oss/capsize/pull/117/commits/fd9a77fd86f8872de291018430cd9e1fc7a4eaf4#diff-8ea7ce0197e120f4013056bd3c56951e98c4ee563aafbf60de94655dad0265d5L9)

**Review by mrm007 — APPROVED** [review]
